server/src/main/java/org/jboss/as/server/deployment/DeploymentUnitPhaseService.java

```diff
@@ -141,10 +140,17 @@ public void handleEvent(final ServiceController<?> controller, final LifecycleEv
             }
         }
 
-        Set<String> subSystemNames = list.stream().map(e -> e.getSubsystemName()).collect(Collectors.toSet());
-        Set<String> registeredSubSystems = (phase == Phase.STRUCTURE) ? new HashSet<>() : deploymentUnit.getAttachment(Attachments.REGISTERED_SUBSYSTEMS);
-        registeredSubSystems.addAll(subSystemNames);
-        deploymentUnit.putAttachment(Attachments.REGISTERED_SUBSYSTEMS, registeredSubSystems);
+        final Set<String> registeredSubSystems;
+        if (phase == Phase.STRUCTURE) {
+            registeredSubSystems = new HashSet<>();
+            deploymentUnit.putAttachment(Attachments.REGISTERED_SUBSYSTEMS, registeredSubSystems);
```

Comment:
The code is correct but I find the new sequence hard to read: we instantiate the `registeredSubSystems` and put it as an attachment. And then, *later* we fill registeredSubSystems with the names of the subsystems...

````
Set<String> registeredSubSystems = (phase == Phase.STRUCTURE) ? new HashSet<>() : deploymentUnit.getAttachment(Attachments.REGISTERED_SUBSYSTEMS);
for (...) {
}
deploymentUnit.putAttachment(Attachments.REGISTERED_SUBSYSTEMS, registeredSubSystems);
````

Reply:
That was to remove a redundant put which I believe will mean the underlying collection of attachments will need to compare the keys of the attachments present for every phase except for Structure for each deployment being deployed.


Alternatively the if would need to be repeated just for the purpose of separating out the putAttachment to a later point.